cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+		}
+
+		sroot, err := cid.Decode(cctx.Args().First())
+		if err != nil {
+			return fmt.Errorf("failed to parse input: %w", err)
+		}
+
+		fsrepo, err := repo.NewFS(cctx.String("repo"))
+		if err != nil {
+			return err
+		}
+
+		lkrepo, err := fsrepo.Lock(repo.FullNode)
+		if err != nil {
+			return err
+		}
+
+		defer lkrepo.Close() //nolint:errcheck
+
+		bs, err := lkrepo.Blockstore(ctx, repo.UniversalBlockstore)
+		if err != nil {
+			return fmt.Errorf("failed to open blockstore: %w", err)
+		}
+
+		defer func() {
+			if c, ok := bs.(io.Closer); ok {
+				if err := c.Close(); err != nil {
+					log.Warnf("failed to close blockstore: %s", err)
+				}
+			}
+		}()
+
+		mds, err := lkrepo.Datastore(context.Background(), "/metadata")
+		if err != nil {
+			return err
+		}
+
+		cs := store.NewChainStore(bs, bs, mds, filcns.Weight, nil)
+		defer cs.Close() //nolint:errcheck
+
+		cst := cbor.NewCborStore(bs)
+		store := adt.WrapStore(ctx, cst)
+
+		st, err := state.LoadStateTree(cst, sroot)
+		if err != nil {
+			return err
+		}
+
+		height, err := strconv.Atoi(cctx.Args().Get(1))
+		if err != nil {
+			return err
+		}
+
+		//get all the votes' signer ID address && their vote
+		vj, err := homedir.Expand(cctx.Args().Get(2))
+		if err != nil {
+			return xerrors.Errorf("fail to get votes json")
+		}
+		votes, err := getVotesMap(vj, st)
+		if err != nil {
+			return xerrors.Errorf("failed to get voters: ", err)
+		}
+
+		type minerBriefInfo struct {
+			rawBytePower abi.StoragePower
+			dealPower    abi.StoragePower
+			balance      abi.TokenAmount
+		}
+
+		// power actor
+		pa, err := st.GetActor(power.Address)
+		if err != nil {
+			return xerrors.Errorf("failed to get power actor: \n", err)
+		}
+
+		powerState, err := power.Load(store, pa)
+		if err != nil {
+			return xerrors.Errorf("failed to get power state: \n", err)
+		}
+
+		//market actor
+		ma, err := st.GetActor(market.Address)
+		if err != nil {
+			return xerrors.Errorf("fail to get market actor: ", err)
+		}
+
+		marketState, err := market.Load(store, ma)
+		if err != nil {
+			return xerrors.Errorf("fail to load market state: ", err)
+		}
+
+		//init actor
+		ia, err := st.GetActor(_init.Address)
+		if err != nil {
+			return xerrors.Errorf("fail to get init actor: ", err)
+		}
+
+		initState, err := _init.Load(store, ia)
+		if err != nil {
+			return xerrors.Errorf("fail to load init state: ", err)
+		}
+
+		initMap, err := initState.AddressMap()
+		if err != nil {
+			return xerrors.Errorf("fail to load init map: ", err)
+		}
+
+		lookupId := func(addr address.Address) address.Address {
+			if addr.Protocol() == address.ID {
+				return addr
+			}
+
+			var actorID cbg.CborInt
+			if found, err := initMap.Get(abi.AddrKey(addr), &actorID); err != nil {
+				panic(err)
+			} else if found {
+				// Reconstruct address from the ActorID.
+				idAddr, err := address.NewIDAddress(uint64(actorID))
+				if err != nil {
+					panic(err)
+				}
+				return idAddr
```

Comment:
why are we doing this?
seems redundant/unnecessary when we have https://github.com/filecoin-project/lotus/blob/d11248b9989174599e7472306f3f99cd6f8d68b9/chain/state/statetree.go#L310

Reply:
I mean https://github.com/filecoin-project/lotus/blob/d11248b9989174599e7472306f3f99cd6f8d68b9/chain/state/statetree.go#L332

Reply:
Totally right (i do want a version of this that panics instead of erroring, but it can be a one-liner. Thank you!